lib/doorkeeper/models/application_mixin.rb

```diff
@@ -49,5 +49,14 @@ def by_uid(uid)
     def redirect_uri=(uris)
       super(uris.is_a?(Array) ? uris.join("\n") : uris)
     end
+
+    ##
+    # Securely compare the given input secret with
+    # this application's (optionally hashed) secret.
+    # To fall back to older plaintext applications, also compares plain values.
+    def secret_matches?(input)
+      secret == Doorkeeper.configuration.hashed_or_plain_token(input) ||
+        secret == input
```

Comment:
I realize now that if we make `hash_application_secrets` configurable you'll need to expose an option block for handling verification:

```ruby
verify_application_secrets do |ciphertext, input|
  BCrypt::Password.new(ciphertext) == input
end
```

```ruby
def secret_matches?(input)
  verify_application_secrets.call(secret, input) || secret == input
end
```

Reply:
> Technically this could cause a problem where the hashed secret of one application collides with the plaintext of another presenting a security issue, but the odds of that are so ridiculously low it's barely worth mentioning.

Indeed. I suggest we add a note to the configuration that a cryptographically secure, collision-resistant one-way function is mandatory. I did not want to coerce a specific digest mechanism.

Reply:
I can follow your argument regarding application secrets, and while bruteforcing even unsalted SHA256 is and will remain unfeasible for quite some time, I don't mind using a split method for either.

Maybe we should not expose the actual hashing function but simply choose SHA256 and BCrypt ? It would reduce the error potential in users to choose a wrong or insecure method or implement it wrongly. /cc @nbulaj  what do you think?

Reply:
There's an expression: _"Give 'em enough rope, and they'll hang themselves"_

I'm comfortable with making Doorkeeper decide these secure hashing algorithms instead of leaving it to a developer who, while well intentioned, makes a critical error.

We can always consider introducing configuration if people raise an issue in the future.

Reply:
I think it would be great to propose a more secured and stable solution by default, but definetly with the ability to reconfigure it (for those developers that know what they are doing). WDYT?

Reply:
From a security standpoint secure and sane defaults is just as good as no configuration at all. I leave it up to y'all to decide what's right for Doorkeeper.

A good comment explaining rationale for why bcrypt and sha256 goes a long way should we opt for exposing configuration.

Reply:
I had a lot of cases (both my own and my colleagues too) when we need to patch default gem behavior just because it couldn't be configured and doesn't suite our (customer) needs (maybe not connected with hashing algorithms, but ...), so I prefer to have default behavior **with the ability** to change it (instead of monkey-patching :see_no_evil:  ). Also in such case everybody could migrate to any hashing solution in case when Bcrypt (%`your_solution`%) would be easy to brute-force (without need to update to a new Doorkeeper release especially if it breaks existing API).

Reply:
The problem with the configuration option is that we need that twice if we're going for different variants for application and token secrets.

My suggestion is we have a setter on the `Hashable` mixin that allows implementing classes to define a secret hashing and verification function. It will use `SHA256.hexdigest` by default and will be overridden by `Application` to use BCrypt.

In the same fashion, an experienced user can override these sensible defaults by e.g., using refinements or simply by setting 

```
      self.secret_hash_function = lambda do |plain_token|
        BCrypt::Password.create(plain_token.to_s)
      end
```

without the need to check for any other configuration options (such as whether hashing is enabled for this class). My code will respect that defined method regardless whether it comes from the class or another user.

It shouldn't be too easy to override these methods because:

1. The user must be absolutely certain their implementation is correct
2. They must never change this hashing method without additional migration efforts. Putting a configuration option **suggests** that it can be turned on and off as they like.